### PR TITLE
Table functions information_schema_schemata() and information_schema_tables()

### PR DESCRIPTION
**Pull request by jwills**
Per issue #658, we would like to have support for the metadata tables defined in the `information_schema` schema (specifically, `schemata`, `tables`, and `columns`) to support tools like DBeaver and dbt that depend on this metadata in order to operate against a broad class of SQL databases. This PR adds the `information_schema_schemata()` and `information_schema_tables()` table functions that implement the columns in [information_schema.schemata](https://www.postgresql.org/docs/9.1/infoschema-schemata.html) and [information_schema.tables](https://www.postgresql.org/docs/9.1/infoschema-tables.html) in support of this goal.

Since this is my first PR to the project (and first time writing C++ in a pretty long time), I thought I would submit this bit to get feedback on while I work on implementing the more involved (but still straightforward) `information_schema_columns()` table function in a separate PR.

### CODE REVIEW

File: src/function/table/information_schema/information_schema_schemata.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+	// start returning values
+	// either fill up the chunk or return all the remaining columns
+	for (idx_t i = data.offset; i < next; i++) {
+		auto index = i - data.offset;
+		auto &entry = data.entries[i];
+
+		// return values:
+		// "catalog_name", PhysicalType::VARCHAR
+		// TODO(jwills): how to determine this?
```
Comment (hannesmuehleisen) [inline src/function/table/information_schema/information_schema_schemata.cpp:83]: are you looking for the catalog name? its in entry->name no?

Comment (hannesmuehleisen) [reply src/function/table/information_schema/information_schema_schemata.cpp:83]: Ah no there is only one catalog indeed. So "main" is fine, NULL would also be an option. Depends on how tools handle this I think.

**Comment by hannesmuehleisen** [comment]
Thanks for this PR, looks great overall!

File: src/function/table/information_schema/information_schema_schemata.cpp (hunk 1 above)
Comment (jwills) [reply src/function/table/information_schema/information_schema_schemata.cpp:83]: In postgres it is ["always the current database"](https://www.postgresql.org/docs/9.1/infoschema-schemata.html), which I think is equivalent to "main" from a DuckDB perspective?

**Comment by jwills** [comment]
> Thanks for this PR, looks great overall! One thought that I had was that the two and soon three physical functions looked quite similar. Perhaps that could be unified a bit?

Totally-- what did you have in mind? Consolidating them into a single .cpp file? Or something more involved?

File: src/function/table/information_schema/information_schema_schemata.cpp (hunk 1 above)
Comment (Mytherin) [reply src/function/table/information_schema/information_schema_schemata.cpp:83]: "main" is the name of the primary schema, I think setting this to NULL makes more sense. Either that, or setting it to the current database file, as is the case in `pragma_database_list`.

**Comment by Mytherin** [comment]
> Thanks for this PR, looks great overall! One thought that I had was that the two and soon three physical functions looked quite similar. Perhaps that could be unified a bit?

Looking at these functions indeed they look very similar, but I don't think there is a nice way of unifying them. The overlap is relatively small. The column names are all different, the entries which are scanned are different (e.g. _schemata scans the schemas, _tables scans all the tables), and the column names which are extracted are all different. The only part that really could be unified is the loop iteration code:

```cpp
	if (data.offset >= data.entries.size()) {
		// finished returning values
		return;
	}
	idx_t next = min(data.offset + STANDARD_VECTOR_SIZE, (idx_t)data.entries.size());
	output.SetCardinality(next - data.offset);
```

I think that's too small to be worth the trouble of unifying.

**Comment by hannesmuehleisen** [comment]
great, so then I think the only thing missing is to create `information_schema` with some views on startup. Do you need any help there?

File: src/function/table/information_schema/information_schema_schemata.cpp (hunk 1 above)
Comment (jwills) [reply src/function/table/information_schema/information_schema_schemata.cpp:83]: @Mytherin my read from [pragma_database_list](https://github.com/cwida/duckdb/blob/master/test/sql/pragma/test_pragma_database_list.test) was that the name of the database is always "main".

For my use case, the value of this field doesn't actually matter, so if the consensus is for NULL, I am 👍 to make that change.

**Comment by jwills** [comment]
> great, so then I think the only thing missing is to create `information_schema` with some views on startup. Do you need any help there?

Yes, if you're ready for that on this PR, just tell me where that should go. I assume that it needs to be a "special" schema (similar to `TEMP_SCHEMA`) in that a) it can't be dropped and b) the `information_schema` and its tables should not be included in the results of these table generating functions.

**Comment by Mytherin** [comment]
Indeed, I think we should add a new flag to catalog entries `bool is_built_in` or so, which is set for built-in objects and checked before dropping to ensure that the built-in catalog views/schemas/etc cannot be dropped. Right now I believe we do a hard-coded comparison on the schema name, but generalizing this seems cleaner and a better way to go. The initialization of the information_schema can go in StorageManager::Initialize for now.

**Comment by jwills** [comment]
So, it turns out (much to my surprise!) that in Postgres, you *can* drop the `information_schema` and all associated views via a cascade, if you're so inclined! Will update the PR to create the schema and the associated views in initialization and filter them out from the data we show when we query the table functions, but will otherwise leave the `is_built_in` code for another PR, whenever it makes sense.

**Comment by jwills** [comment]
Apologies y'all, another question: in `StorageManager::Initialize`, what is the best way to create the equivalent of `create view information_schema.schemata as select * from information_schema_schemata()`? I see the example in that block of how to create a schema using `CreateSchemaInfo`, but it seems like the requirements for setting up a `CreateViewInfo` struct are more involved. What's the best way to do it, or is there an example I can borrow?

**Comment by Mytherin** [comment]
For CREATE VIEW indeed you need to create a query node. Creating that by hand is a bit ugly; it's easier to use the parser to create it. You can just put this in the StorageManager::Initialize, after the `context.transaction.Commit();`:

```sql
context.Query("CREATE VIEW sqlite_master AS SELECT * FROM sqlite_master()", false);
```

**Comment by jwills** [comment]
> For CREATE VIEW indeed you need to create a query node. Creating that by hand is a bit ugly; it's easier to use the parser to create it. You can just put this in the StorageManager::Initialize, after the `context.transaction.Commit();`:
> 
> ```sql
> context.Query("CREATE VIEW sqlite_master AS SELECT * FROM sqlite_master()", false);
> ```

Wrote this up as a commit ^^, but it broke more of the unit tests than I expected- essentially anything involving a load/store of a database will fail b/c the information_schema views defined here already exist.

**Comment by Mytherin** [comment]
Yeah that makes sense, this is another reason why these views need to have this `built_in` property: that way we can also avoid writing built-in views to disk. I think the PR as-is is fine and we can add the views later, this is not super critical.

**Comment by jwills** [comment]
Yep, makes sense re: `built_in`. Did a force push to remove the commit where I defined the views; please take another look.

**Comment by Mytherin** [comment]
Looks good to me, thanks for the contribution!

**Pull request merged**